### PR TITLE
Settle the ParsedResult information model before the port (#391)

### PR DESCRIPTION
**Pull request by tylervick**
Documentation only. Settles the information model for #391 before any code exists, and propagates the consequences through the plan. **No implementation** — nothing under `Sources/` or `Tests/` changes.

## What needs review: the eight decisions

`ParsedResult` is the artifact this migration would otherwise build twice. Shaped so today's templates render unchanged, it isn't backend-neutral — it's legacy-shaped, and the modern reader spends its life supplying `nil` for fields that exist only because the old UI reads them. Task 2.5's recommendations, applied as written:

| # | Question | Answer | Why |
|---|---|---|---|
| 1 | Per-activity durations? | **No** — drop `finish` | Modern publishes `startTime` only. No duration beats a fabricated `(0.00s)`. |
| 2 | Five activity-type states? | **No** — drop `activityType`, keep `isFailure` | The one useful state (`userCreated`) has no modern source at any fidelity. |
| 3 | `ObjectClass` in the model? | **No** — delete it | `IDESchemeActionTestSummaryGroup` is an Xcode internal rendered into a CSS class. |
| 4 | Attachment UTI as a field? | **No** — `filenameExtension` only | `AttachmentType` needs only enough to pick a template and a MIME type. |
| 5 | Status as a legacy raw string? | **No** — neutral enum | A raw string forces the modern reader to emit legacy spellings it never saw. |
| 6 | Swift Testing `Arguments`? | **Yes, now** | The only addition *inside* the tree; adding the slot later reshapes the port. |
| 7 | Insights / metrics? | **No, not now** | Top-level and additive later. Cheap to defer, no empty slots. |
| 8 | Duration sums repetitions? | **Yes** — keep today's behaviour | Modern's own node value isn't the sum; summing makes both backends agree. |

I agreed with all eight, so nothing was applied over an objection.

**The direction is the point.** Each answer removes a field from the port or an entry from the differential allow-list. Holding the legacy backend down to the modern backend's capability makes the two agree *by construction* rather than by mask — and the masked region is exactly where a regression hides. The allow-list drops from **five entries to three**: `activityTypeClasses` and `durations` are gone because there is no longer a field to diverge. The cost is honest and one-way: legacy stops rendering things it could have. That's a 4.0 behaviour change, made once and visible in the model.

## Two constraints found while applying, not assumed away

**Answer 6 is unexercised by any fixture.** `Arguments` is in the published `TestNodeType` enum (`xcresulttool get test-results tests --schema`), but `SwiftTestingSuite` has no parameterized case and all three bundles contain **zero** `Arguments` nodes. Adding the field on the strength of a schema is defensible — it's a tree reshape, expensive later — but shipping a field nothing populates and nothing checks is not. Task 8 gains a step that adds a parameterized `@Test` to the sample app, and its test **fails until that lands**, with an explicit instruction to remove the field and revisit answer 6 if the nodes don't appear.

**Answer 4 can't use `UTType(filenameExtension:)`** — macOS 11+ against a 10.15 floor. The mapping keeps an explicit table, matching how `Attachment.swift` already guards `UTType` for MIME types.

## Two things the rebase exposed

`origin/main` brought in #430 (deterministic report identifiers), and Task 5 still said `Activity.uuid` becomes `UUID().uuidString`. That would compile and **silently undo #430** — `ReproducibilityTests.testRenderingTheSameBundleTwiceProducesIdenticalBytes` asserts byte-identical renders, and a random id per render breaks it immediately. `Activity` now mints from `IdentifierPath` like every other model, which also satisfies the rule this task exists to enforce: a path-derived id is backend-neutral, where an activity uuid would have been legacy-only.

Removing the activity-type and duration fields orphans two template placeholders. They're fed empty strings rather than deleted, because **`createTemplates.sh` — the generator named in `HTMLTemplates.swift`'s own DO-NOT-EDIT header — is not in this repository.** That leaves a cosmetic empty paren on activity rows (`Some activity ()`), which I've routed to the redesign workstream rather than hand-editing generated code. Still better than `(0.00s)`: one is untidy, the other is a fabricated number.

## Everything else is mechanical propagation

Task 3's model, Task 4 (legacy reader), Tasks 8–10 (modern reader), Task 5 (renderer), Task 12 (allow-list and masker), plus the spec's "not a superset" table, where voided rows now read **dropped** (a decision) rather than **lost** (an asymmetry). **Task 10 is reframed** — attachment typing was written as a modern-only workaround; under answer 4 it's how *both* backends type, so it gains a cross-backend agreement test.

Verified mechanically rather than by eye: every `ParsedActivity`/`ParsedIteration`/`ParsedTestCase`/`ParsedAttachment` construction in the plan matches the amended shape, the allow-list JSON parses, and its three rules have exactly three masker implementations with no orphans either way.

## What happens after this merges

This merges **documentation**. Implementation starts after, following the repo's existing convention: one worktree per phase, branches as `tylervick/<slug>-<issue>`, one PR per phase, `main` green throughout. Task 2.5 is complete and its steps are checked off; Task 3 is the first code.

## Verification

Rebased onto `origin/main` (was 4 behind, clean — the branch touches only `docs/`). The plan's own Task 1 Step 0 now passes: `IdentifierPath.swift` present, `ReproducibilityTests` green at 5 tests / 0 failures.

Full suite after regenerating fixtures: **28 tests, 1 skipped, 0 failures**, exit 0.

The regeneration was necessary, and the reason is worth recording. Before it, three tests failed — `testAttachmentsExist`, `testDownsizedAttachmentsExist`, `testResultStatusCount` (`16` vs `17`). Nothing to do with this diff: #428 added `FirstSuite.testAttachScreenshot` to the sample app at 00:10, and my bundles were generated at 22:16 the night before. Stale fixtures, not a regression.

One self-inflicted note: my first run reported `EXIT=0` while carrying those three failures, because it piped through `tail` and read `tail`'s status. That is exactly the trap this plan documents as a Global Constraint — I wrote the warning and then walked straight into it, which is a reasonable argument that the constraint earns its place. The verified run above uses `set -o pipefail`.

Refs #391. Milestone 4.0.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Added a design specification for migrating result processing from legacy tooling to modern and legacy-compatible formats.
  * Documented backend selection, status handling, repetitions, failures, attachments, and fault behavior.
  * Defined a replacement JSON output schema and clarified compatibility expectations.
  * Added plans for differential testing, deterministic identifiers, CI coverage, phased implementation, and migration risks.
  * Clarified scope and future considerations for result-processing improvements.

### CODE REVIEW

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+| log section `emittedOutput` | absent; structured `messages` instead | **lost** |
+| `exportRecursiveJson()` (drives `--json`) | no equivalent | **no replacement** |
+| repetitions (duplicate siblings + `repetitionPolicySummary`) | first-class `nodeType: "Repetition"` | **improved** |
+| Swift Testing `Arguments` / `Expression` / `Test Value` | first-class node types | **modern-only** (answer 6) — slot added now, unexercised by fixtures |
+| tree shape | two fewer wrapper levels | differs, see below |
+
+Consequence: byte-identical output across the two backends is not achievable.
+The bar is a *declared and reviewed* diff, asserted in CI, not an empty one —
+though the model decisions shrink what has to be declared from five entries to
+three, which makes the remaining diff a stronger proof than a larger masked one.
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md]: Fixed. The spec now names all three with what still differs and which fixture exercises each:

| Rule | What still differs | Exercised by |
|---|---|---|
| `attachmentDisplayNames` | Modern exposes only the generated filename, so names fall back to the type-derived label | `TestResults` — `testWithSpecialChars()`, `testAttachHtmlData()`, both of which set `XCTAttachment.name` |
| `failureTitlePrefix` | Legacy renders `<issueType> at <file>:<line>:<message>`; modern gives `<file>:<line>: <message>` pre-joined, no `issueType` | `TestResults` every failing case; `RetryResults` `testJustFail()` |
| `wrapperGroups` | Modern omits `Selected tests` / `All tests` / `<target>.xctest` — a deliberate render difference, not a format loss | both bundles |

Also stated that everything outside these three must be byte-identical after masking, and that the preferred response to a new divergence is deleting the field from the port rather than adding a row — which is what answers 1, 2 and 4 did.

Worth noting the spec and plan now agree mechanically: the three names in the spec table, the three `rule` values in `differential-allowlist.json`, and the three entries in `KnownLossMasker.implementedRules` are the same set, verified rather than eyeballed.

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+
+An explicit override is **required**, not optional: the differential tests must
+force each reader on the same bundle regardless of what the host toolchain
+advertises. Exposed as a CLI option (`--result-reader legacy|modern|auto`,
+default `auto`) so the harness can drive it through the existing
+`xchtmlreportCmd` helper.
+
+Any hard failure of a legacy command also demotes the backend to modern, so a
+version string that changes shape unexpectedly degrades to working rather than
+broken.
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md]: Good catch, and it was a spec/plan inconsistency as well as a hazard.

`ResultBackend.resolved()` in the plan only ever consults `legacyCommandsAvailable` — it never demoted on arbitrary command failure. The spec sentence claimed more than the implementation did, and what it claimed was wrong: a corrupt bundle, a permission error, or a truncated `.xcresult` would have been treated as evidence that legacy support is gone, silently retried on the modern reader, and returned a partial report where a clear failure belonged.

Demotion is now scoped to capability detection only, with the negative case spelled out explicitly so the broader reading cannot creep back in. Everything else propagates through `FaultCollector` (`.missingInvocationRecord`, `.missingActivities`, `.payloadExportFailed`) to the exit-3 path, exactly as it would if only one backend existed.

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+`FirstSuite.swift:66: ...` — is a declared entry in the diff allow-list.
+
+## `--json` becomes our own schema
+
+`--json` is currently `exportRecursiveJson()`: a raw dump of Apple's internal
+legacy object graph. It has no new-format equivalent.
+
+`--json` will emit the `ParsedResult` model as JSON, with a documented schema,
+**identically on both backends**. This is a breaking change to `--json` output
+for every user, landing in 4.0 with release notes.
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md]: Agreed, and taken — though as a requirement on the task that owns `--json` rather than as a contract authored here, since the final `ParsedResult` shape is what it has to describe and this PR is what settles that shape.

You are right about the underlying problem: `ParsedResult` is an internal Swift model, so deriving public output from a synthesized `Encodable` publishes whatever the type happens to look like on the day, and every later field rename becomes a silent breaking change for consumers.

Task 14 now has a step *before* the encoder step requiring `docs/json-schema.md`, covering the specific gaps you named: field names and nesting with a worked example from `SanityResults`; `ParsedStatus` as lowercase strings rather than ordinals; one uniform null-versus-omitted rule; durations as seconds and timestamps as ISO-8601; array ordering guarantees so two reports can be diffed; and a top-level `schemaVersion` with a stated bump policy. The instruction is explicit that the encoder is made to match the document, not the reverse.

Flagged in the spec as a Task 14 deliverable rather than a follow-up — the flag is public output and #391 is where it changes, so deferring it past the break would be the wrong order.

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+An explicit override is **required**, not optional: the differential tests must
+force each reader on the same bundle regardless of what the host toolchain
+advertises. Exposed as a CLI option (`--result-reader legacy|modern|auto`,
+default `auto`) so the harness can drive it through the existing
+`xchtmlreportCmd` helper.
+
+Demotion is scoped to **capability detection only**. `auto` and an explicit
+`legacy` both fall back to `modern` when the toolchain does not advertise the
+legacy commands, so a version string that changes shape unexpectedly degrades
+to working rather than broken.
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md:242]: You are right, and this one was mine — I introduced the substitution in the previous round while fixing a different finding, and it created a worse problem than the one it solved.

The hazard is exactly as you describe: `DifferentialTests` forces each reader, so on a modern-only host an explicit `legacy` that quietly became `modern` would compare the modern backend against itself and report parity. A differential that cannot fail is the precise failure mode this suite exists to prevent.

Detection is now tri-state and only `auto` ever substitutes:

| Requested | `available` | `unknown` | `unavailable` |
|---|---|---|---|
| `auto` | legacy | **modern** | modern |
| `legacy` | legacy | **legacy** | **error, no substitution** |
| `modern` | modern | modern | modern |

`resolve()` returns `.legacyUnavailable` rather than a substituted backend, and the CLI turns that into a `ValidationError`. Also took your point about the unparseable version string: that is now `unknown` rather than proof of absence, so `auto` degrades to modern (working beats broken) while an explicit `legacy` still attempts legacy and lets any real command failure surface as a fault.

And `requireBothBackends()` now asserts on the backend it actually resolved instead of trusting the request, so a future substitution bug fails the guard rather than silently hollowing out every comparison beneath it.

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+
+It is deliberately **not** "any hard failure of a legacy command demotes".
+A corrupt bundle, a permission error, a truncated `.xcresult`, or a subprocess
+that dies for an unrelated reason are not evidence that legacy support is
+gone — and silently retrying them on the modern reader would replace a clear
+failure with a partial report that looks fine. Those failures propagate through
+`FaultCollector` (`.missingInvocationRecord`, `.missingActivities`,
+`.payloadExportFailed`) and reach the exit-3 path, exactly as they would if
+only one backend existed.
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md:251]: Fixed in both, and you were right that it needed to be both.

I had recorded the fault in `ModernPayloadStore.exportLogs` and missed that `ResultFile.exportLogs` has the identical gap — it logs the write failure and returns nil with no fault, so a report missing a log it successfully *read* still exits 0. That one is pre-existing rather than introduced by the port, but Task 5 already modifies that file, and leaving one of two implementations silent is precisely how the asymmetry survives a refactor.

Both `catch` blocks now record `.logExportFailed`, and Task 5 calls for `FaultReportingTests` coverage asserting it appears on each backend when the destination cannot be written.

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+- enum encoding (`ParsedStatus` as lowercase strings, not ordinals)
+- null versus omitted for every optional — one rule, applied uniformly
+- duration units (seconds as a JSON number) and timestamp format
+- array ordering guarantees, so consumers can diff two reports
+- a top-level schema version, and what a consumer should do when it changes
+
+Until that exists the `--json` change is specified only as "our schema", which
+is not a contract anyone can build against. This is a Task 14 deliverable, not
+a follow-up: the flag is public output, and #391 is where it changes.
+
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md:487]: Agreed — as written the acceptance criteria contradicted the allow-list, which deliberately preserves three render-level differences.

Settled as **schema identity, not value identity**. Both backends emit the same field names, nesting, enum encoding and `schemaVersion`; every key present on one is present on the other. Values may differ in exactly four places and nowhere else:

- `attachment.name` — populated on legacy, `null` on modern
- failure activity titles — legacy carries the `<issueType> at ` prefix
- group nesting — legacy has the wrapper levels
- `testCase.arguments` — populated on modern for parameterized Swift Testing cases, always `[]` on legacy

Anything else is a reader bug, not a permitted variance. Good catch on `arguments` in particular: I added that field in this PR and had not carried it into the `--json` reasoning.

The test was weak in the same direction — it compared `Set(keys)` at the top level only, which cannot see a nested field present on one backend and absent on the other. It now compares full key paths and asserts matching `schemaVersion`.

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+- `attachment.name` — populated on legacy, `null` on modern
+- failure activity titles — legacy carries the `<issueType> at ` prefix
+- group nesting — legacy has the wrapper levels, modern does not
+- `testCase.arguments` — populated on modern for parameterized Swift Testing
+  cases, always `[]` on legacy
+
+Any other value difference is a bug in a reader, not a permitted variance.
+Stating this matters because the acceptance criteria would otherwise contradict
+the allow-list, which deliberately preserves three render-level differences.
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md]: Right — four bullets under a heading that says three entries remain. `arguments` was sitting there as an implicit fourth allow-list rule, which is exactly the kind of quiet expansion the allow-list is supposed to make impossible.

It is not an allow-list entry, and now says so explicitly. The two classes are separated:

- **Class 1**, the three differential rules (`attachmentDisplayNames`, `failureTitlePrefix`, `wrapperGroups`) — render-level differences that surface in `--json` for the same reason they surface in HTML.
- **Class 2**, `testCase.arguments` — a model-level capability difference. It needs no masking rule because **nothing renders it**: the templates have no slot for arguments, so it cannot appear in the HTML differential at all. It is visible only in `--json`.

To your question about parameterized cases: the differential does not exclude them, it simply never sees this field. `--json` comparison must handle `arguments` separately — asserting legacy is `[]` rather than asserting the two sides are equal.

I also added the fixture limitation that falls out of it, since it is the more dangerous half: until the parameterized `@Test` lands in Task 8, no bundle produces a non-empty `arguments` on either backend, so a naive equality assertion passes with both sides `[]` and proves nothing about the modern reader.

File: docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+4.0 with release notes.
+
+**"Identical across backends" means schema identity, not value identity.** Both
+backends emit the same field names, nesting, enum encoding, and
+`schemaVersion`; every key present on one is present on the other. The *values*
+may differ exactly where the differential allow-list already says they differ,
+and nowhere else:
+
+- `attachment.name` — populated on legacy, `null` on modern
+- failure activity titles — legacy carries the `<issueType> at ` prefix
```
[the remark this answers is not in the document]
Comment (tylervick) [reply docs/superpowers/specs/2026-08-10-xcresulttool-legacy-migration-design.md]: Fixed — that span is gone entirely in the rewrite, which now uses `<issueType>` without the trailing text inside the backticks. Swept both documents for MD038 (code spans with leading or trailing whitespace, outside fenced blocks): zero remaining in either.

**Comment by tylervick** [comment]
Worked the xhigh review. Both wrong decisions corrected and the propagation re-derived rather than patched; Task 5 split; all 15 findings addressed. Local suite green (28 tests, 1 skipped, 0 failures).

Everything factual below I re-ran against the committed fixtures on Xcode 26.2 rather than taking from the brief.

## The two decisions

**Decision 3 → replace, not delete.** Confirmed, and the blast radius is wider than reported. `ITEM_CLASS` empty renders `<div class=" failed">`, which breaks:

- `showElementsWithSelector('.run.active .test-summary.succeeded')` and its `.skipped` / `.failed` / `.mixed` siblings — the "show only failures" filter matches nothing
- `querySelectorAll('.run.active .test-summary-group')` and the `classList.contains('test-summary-group')` filter under it — group collapse
- stylesheet rules on both classes (`.test-summary p`, `.test-summary-group > p`, `.test-summary.<status> .test-result-icon`, …)
- **four** test call sites, not one: `CoreTests.swift:122`, and `ReproducibilityTests.swift` 138, 174, 175

The raw `IDESchemeActionTest*` values were the legacy part and still go. The emitted class names are the report's own contract and stay, behind a renderer-side `NodeKind`. The port needs no field — `ParsedNode` already distinguishes group from case — so this also keeps the "no legacy shape in the model" rule intact.

**Decision 1 → field still goes, two corrections.** `start` is now named as the replacement ordering key, in the spec and in Task 4, with the invariant spelled out: that sort interleaves failure rows among activities so a failure renders where it occurred.

`durations` is back on the allow-list. Verified across all three fixtures — `durationInSeconds` is `null` on **every** `Test Suite`, `Test Plan`, `UI test bundle` and `Unit test bundle` node, against real legacy values:

| Group | Legacy | Modern |
|---|---|---|
| `FirstSuite` | 0.699s | `null` → `0.00s` |
| `SecondSuite` | 0.126s | `null` → `0.00s` |
| `ThirdSuite` | 0.132s | `null` → `0.00s` |
| `SampleAppUnitTests` | 0.213s | `null` → `0.00s` |

Swift Testing test *cases* are null the same way. `wrapperGroups` only drops lines containing `Selected tests`/`All tests`/`.xctest`, so these were diverging unmasked. Allow-list is **four** entries; every "three" updated in both documents.

## Task 5 split

5a moves the renderer onto `ParsedResult` and must be byte-identical against the Task 2 baseline. 5b applies the decisions that change output, gated on an enumerated diff of exactly three shapes (activity class loses its type token, activity row loses `(1.23s)`, activity id becomes a path digest). The spec's phase 1 split the same way.

`ObjectClass → NodeKind` moved *into* 5a, since it now preserves bytes. `Activity.uuid → IdentifierPath` moved into 5b, since it changes them.

## Findings 5–15

All addressed. Notable ones: the failure double-count is now deduped on the message tail (both documents describe the same failure — activities gives `XCTAssertTrue failed - Test failed`, tests gives `FirstSuite.swift:86: XCTAssertTrue failed - Test failed`); `failureTitlePrefix` strips both shapes; `PayloadProviding` gains `url` plus the two content helpers as a protocol extension, with a note that keeping `ResultFile` at those sites to get a build would silently vanish screenshots and videos; the three XCResultKit accessors stay on `ResultFile`; `Summary.init` keeps its `resultIndex`/`actionIndex` seeding.

## Two things I did beyond the brief

**The `durations` mask is necessarily over-broad, and that costs coverage.** The duration sits on a different line from the `test-summary-group` class, so it cannot be scoped by line — normalising all `(N.NNs)` therefore also hides XCTest case durations, which *do* agree. Added `testXCTestCaseDurationsAgreeAcrossBackends`, asserting on the model, to restore exactly that coverage, with both sides commented not to delete one without the other.

**Finding 8's rule applies to the legacy reader too.** The brief asked for a stated rule covering every reader, and fixing only `ModernResultReader` would have left the identical hole: `LegacyResultReader.read()` returns `ParsedResult(runs: [])` when no `ActionRecord` parses, which satisfies `Summary.init`'s guard, records no fault, and exits 0. Both readers now return nil, and the spec states it as a contract rather than a case fix.

## Disagreements

None on the findings themselves — all 15 reproduced. One qualification on finding 7: the per-run split is now structured (`testables(on:from:)`) but still returns the whole tree, because no fixture has two destinations and `Device` is itself a `TestNodeType`, so the nested shape is unverified. I have written that in the comment rather than implementing a filter I cannot test. Implementing it blind would be worse than the current honest placeholder — but it does mean finding 7 is *structurally* addressed rather than *behaviourally* fixed, and it should not be marked resolved without a two-destination bundle.
